### ISSUE TITLE
Null items in collections mapping

### ISSUE THREAD

**Issue by Chuckame** [mapstruct/mapstruct#2684]
# Context
Collection mapping output can contains nulls if the corresponding mapping method returns null.

To prevent having null items, 2 solutions:
- Having an after method that is removing all null items
- implement a method that is taking a list, mapping item per item, filter null items

# Non goal
Complex filtering with custom classes and/or annotations

# Proposal
I saw multiple null strategies.

We can add an collectionNullItemMappingStrategy allowing filtering output to only get non-null items into generated collections

**Comment by filiphr** [comment]
Your request makes a lot of sense. I would even argue that this should be done out-of-the-box without a `collectionNullItemMappingStrategy`, since someone expecting to have `null`(s) in a collection is a code smell in my personal opinion.

This can actually be implemented in multiple steps. 

1. Explicitly do null checks before adding to a collection. Similar to how we do bean mappings
2. Use what is proposed in #2610 and #1610 to achieve the non goal through using the new `@Condition`.

**Comment by Chuckame** [comment]
I totally agree with you on a point : null items should not be added to a collection since mapstruct is nullsafe-like, but there is a problem if we don't add them: how to deal with simple collection mapping that contains null elements? I think some teams don't want to lose data, and other don't want null pointer... 😅 This is why I proposed to simply add this param

**Comment by ytqxyd** [comment]
Thank you! I was trying to figure out why nullValuePropertyMappingStrategy doesn't work on a collection and I found this open ticket. 

In the meantime, is there any workaround such as java expression or default mapper?

**Comment by teamsameem** [comment]
@filiphr  is there any workaround as we are currently blocked on this

**Comment by filiphr** [comment]
@teamsameem you can write a custom method for mapping your iterables.

e.g.

```java
@Mapper
public interface MyMapper {
    Target map(Source source);

    default List<Target> map(List<Source> list) {
        if ( list == null ) {
            return null;
        }

        List<Target> result = new ArrayList<>( list.size() );

        for(Source source: list) {
            if (source == null) {
                // Or any other condition
                continue;
            }
            result.add(map(source));
        }
        return result;
    }
}
```

**Comment by teamsameem** [comment]
have taken exactly the same approach, due to the highly nested natured of the object , its not much of an ease this way

**Comment by frannajaya** [comment]
@teamsameem for me, I add @AfterMapping method for removing the null result of the mapping object
e.g.
```
@AfterMapping
  default void removeNullItem(
      @MappingTarget final List<Target> listTarget
  ) {
    listTarget.removeIf(Objects::isNull);
  }
```

**Comment by thunderhook** [comment]
This could also be handled with a condition when we tackle:
* #1610

**Comment by Obolrom** [comment]
@filiphr 
What do you think about the naive implementation of filtering out null values?
What if we add if statement in `IterableMappingMethod.ftl` to check an item for null when iterating over a collection?

```ftl
for ( <@includeModel object=sourceElementType/> ${loopVariableName} : ${sourceParameter.name} ) {
    if ( ${loopVariableName} == null ) {
        continue;
    }
    <@includeModel object=elementAssignment targetBeanName=resultName targetWriteAccessorName="add" targetType=resultElementType/>
}
```
